Makefile.envs

```diff
@@ -1,4 +1,12 @@
-export PYVERSION ?= 3.12.7
+export PYVERSION ?= $(python3 <<EOF
+import sys, tomllib
+with open("pyproject.toml", "rb") as in_file:
+    requires_python = tomllib.load(in_file).get("project", {}).get("requires-python")
+assert requires_python, "requires-python not found in pyproject.toml"
+sys.stdout.write(f"{requires_python}\n")
+EOF
+)
```

Comment:
I don't think heredoc works in defining makefiles. I think you would need to move this into `tools/get_py_version.py` and do:
```
export PYVERSION ?= $(shell python3.12 get_version.py)
```

Reply:
But I'm concerned about the performance implications of this. Before this change, 
```sh
time make -f Makefile.envs .output_vars 
```
reports that it takes 5 ms. After this change it takes 300 ms so 60x longer. My guess is we source this file about 1000 times in our CI, so this could add as much as 5 minutes to our CI time. WDYT @ryanking13?

Reply:
I don't think this is better than the current way. Let's think more about how we can improve our Makefile.